allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -0,0 +1,509 @@
+        Box filtering functions either apply a filter on a set of boxes and return the filtered set.
+        The elements should evaluate to one of the following:
+            box_filtering_function(set_to_filter, attribute_function, constant)
+            all_boxes
+        TODO(pradeep): We may want to change the order of arguments here to make decoding easier.
+        """
+        if sub_expression[0].startswith('filter_'):
+            # filter_* functions are box filtering functions, and have a lambda expression as the
+            # second argument. We'll process the whole nested structure of the lambda expression
+            # here.
+            function = getattr(self, sub_expression[0])
+            arguments = sub_expression[1]
+            set_to_filter = self._execute_box_filter(arguments[0])
+            flattened_lambda_terms = self._flatten(arguments[1:-1], [])
+            attribute_function = lambda x: self._apply_function_list(flattened_lambda_terms, x)
+            attribute = self._execute_constant(arguments[-1])
+            return function(set_to_filter, attribute_function, attribute)
+        elif sub_expression == 'all_boxes' or sub_expression[0] == 'all_boxes':
+            return self._all_boxes()
+        else:
+            raise RuntimeError("Invalid box filter expression: %s" % sub_expression)
+
+    def _execute_object_filter(self, sub_expression: Union[str, List]) -> Set[Object]:
+        """
+        Object filtering functions should either be a string referring to all objects, or list which
+        executes to a filtering operation.
+        The elements should evaluate to one of the following:
+            object_filtering_function(object_set)
+            negate_filter(object_filtering_function, object_set)
+            all_objects
+        """
+        if sub_expression[0] == "negate_filter":
+            arguments = sub_expression[1]
+            original_filter = getattr(self, arguments[0])
+            initial_set = self._execute_object_filter(arguments[1:])
+            return self.negate_filter(original_filter, initial_set)
+        elif sub_expression == "all_objects" or sub_expression[0] == "all_objects":
+            return self._all_objects()
+        elif isinstance(sub_expression[0], str) and isinstance(sub_expression[1], list):
+            # These are functions like black, square, same_color etc.
+            function = getattr(self, sub_expression[0])
+            arguments = self._execute_object_filter(sub_expression[1])
+            return function(arguments)
+        else:
+            raise RuntimeError("Invalid object filter expression: %s" % sub_expression)
+
+    @staticmethod
+    def _execute_constant(sub_expression: str) -> Union[str, int]:
+        """
+        Acceptable constants are numbers or strings starting with `shape_` or `color_`
+        """
+        if str.isdigit(sub_expression):
+            return int(sub_expression)
+        elif sub_expression.startswith('color_'):
```

Comment:
When would you do this?  Why not just have `blue`, `black`, and `yellow` be the constants?

Reply:
`blue`, `black` and `yellow` are functions. I used `color_` and `shape_` prefixes to differentiate constants from those functions. I'll add a comment about this.